server/src/main/java/io/druid/client/CachingClusteredClient.java

```diff
@@ -205,7 +205,7 @@ public CachingClusteredClient(
 
       // Pull cached segments from cache and remove from set of segments to query
       final Map<Cache.NamedKey, byte[]> cachedValues;
-      if (useCache) {
+      if (useCache && cacheKeys.size() <= cacheConfig.getCacheBulkMergeLimit()) {
```

Comment:
I don't think it should disable populating if it exceeds the size. Future cache fetches may not be requesting the full dataset, and really the background caching should be able to passively do this (though I think there might be buffer corruption risks for the background caching right now)